packages/grid/_modules_/grid/components/AutoSizer.tsx

```diff
@@ -96,13 +96,13 @@ export const AutoSizer = React.forwardRef<HTMLDivElement, AutoSizerProps>(functi
         (!disableHeight && state.height !== newHeight) ||
         (!disableWidth && state.width !== newWidth)
       ) {
-        setState({
-          height: height - paddingTop - paddingBottom,
-          width: width - paddingLeft - paddingRight,
-        });
+        setState(() => ({
```

Comment:
Nice & interesting direction, we push the approach to its full extent in https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/TextareaAutosize/TextareaAutosize.js#L88-L116. We would need to move the onResize callback to an effect, I imagine, if we want to continue in this direction too. Until we do, maybe

```suggestion
        setState({
```

would be simpler?

Reply:
Well by passing a function, we know that it's 'blocking' the UI

Reply:
I don't follow. AFAIK, proving a function or not has no impact on the scheduling of React. I don't see it mentioned in https://reactjs.org/docs/hooks-reference.html#functional-updates nor have I experienced it anytime. 

@eps1lon Is there a scheduling difference between using a setState function setter and providing the state, as is?

Reply:
https://medium.com/@baphemot/understanding-reactjs-setstate-a4640451865b#:~:text=setState%20accepts%20a%20function%20as,Object%20to%20merge%20into%20state.

SetState is async by passing a callback I ensure that it's the latest state that will be rendered

Reply:
This blog post is from 2017, we are no longer using class components.

Reply:
If you depend on previous state then the callback is correct. Otherwise you might set state with stale values.

But that isn't the problem here. The problem is that you have it in an if block that relies on previous state. You need to move the condition inside the callback and return the old state otherwise.

Reply:
> Well by passing a function, we know that it's 'blocking' the UI

What do you mean by that? As far as I know neither overload changes batching semantics.

Reply:
I mean that it will avoid state concurrency issue, it's not blocking but it will guarantee the order which is what I want.

Reply:
> I mean that it will avoid state concurrency issue

That's a fairly generic term. What specific problem are you seeing?

> it will guarantee the order which is what I want.

The overload does not change the order of state updates applied. Maybe you're seeing stale values being set.